### DISCUSSION TITLE
Does this 'link_pre_link_transform' method get called?

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#1145] [Q&A]
See https://github.com/ipspace/netlab/blob/f0d3970d71aeb37bd17e63ebdfdba99c5a82a647/netsim/modules/gateway.py#L111

Everywhere else (vlan, vrf, dhcp) it's called "link_pre_transform", a 'grep' shows no obvious invocations

**Comment by jbemmel** [comment]
I guess it does: https://github.com/ipspace/netlab/blob/f0d3970d71aeb37bd17e63ebdfdba99c5a82a647/netsim/modules/__init__.py#L103

**Comment by ipspace** [comment]
The sequence of hooks called is (hopefully relatively accurately) described here: https://netlab.tools/dev/transform/#global-data-transformation (and subsequent sections).

See also https://netlab.tools/dev/transform/#node-level-module-hooks and https://netlab.tools/dev/transform/#link-level-module-hooks. Those hooks were added to avoid having a "loop over all nodes" or "loop over all links" code all over the place when it's not needed (= when you don't need the global context).

Several modules use the `node_something` hooks (to process data of a single node), not sure any module uses `link_something` hook at the moment.
